File: vgen.js
#!/usr/bin/env node
const program = require('/usr/lib/node_modules/commander');
const StellarSdk = require('/usr/lib/node_modules/stellar-sdk');
const csv = require('/usr/lib/node_modules/fast-csv')
const mongo = require('/usr/lib/node_modules/mongodb').MongoClient;
const assert = require('assert')

program
  .version('0.1')
  .option('-p, --prefix', 'Find address that begins with the words specified')
  .option('-s, --suffix', 'Find address that end with the words specified')
  .parse(process.argv);
suffix = program.suffix || (!program.prefix && !program.suffix);


const url = "mongodb://" + (process.env.MONGO_PASS != null ? process.env.MONGO_PASS + "@" : "") + "localhost:27017";
const nBench = 1000;


VanitySearch = function(words, suffix) {

  console.log(`Finding address ...`);

  mongo.connect(url, { useNewUrlParser: true }, async function(err, db) {
    if (err) throw err;

    var dbo = db.db("mydb");

    var i = oldHit = newHit = 0;
    var oldTime = newTime = process.hrtime();
    do {
      const pair = StellarSdk.Keypair.random();
      const publicKey = pair.publicKey();
      i = i + 1;
      if (i % nBench == 0) {
        newTime = process.hrtime();
        dt = (newTime[0] - oldTime[0]) + (newTime[1] - oldTime[1])*1e-9
        console.error('Generation rate:', Math.round(1.0 * nBench / dt), 'addr/sec\t',
                      'Hit rate: ', Math.round(100*(newHit - oldHit)/dt)/100, 'addr/sec');
        oldTime = newTime;
        oldHit = newHit;
      }
      var word = suffix ? words.find(w => publicKey.endsWith(w)) : words.find(w => publicKey.startsWith(w, 1));
      if (word != null) {
        let r = await dbo.collection("StellarVanityPool").insertOne({ "suffix": word, "public": pair.publicKey(), "secret": pair.secret() });
        assert.equal(1, r.insertedCount);
        newHit = newHit + 1;
      }
    } while (true);
  });
};


console.log(`Reading names...`);
var names = new Array();
csv.fromPath("./names/NAMES.txt", {headers: false, ltrim: true, rtrim: true})
  .on("data", function(data) {
    names.push(data[0]);
  })
  .on("end", function() {
    console.log(`Done.`);

    const words = names.filter((v,i,a) => { return v.length > 2 });

    console.log('Loaded', words.length, 'suffixes.');
    VanitySearch(words, suffix);
  });

File: vgen.py
#!/usr/bin/python3
from stellar_base.keypair import Keypair
from pymongo import MongoClient
from pymongo.errors import ConnectionFailure, OperationFailure
from collections import OrderedDict
import csv
import time
import os
import sys


# how frequently to output performance statistics
NBENCH = 1000


# global variables
keywords = set()
maxlen = 0


# load the list of interesting keywords (suffixes, prefixes)
def load_keywords(filename):

    global maxlen

    iFile = open(filename, 'rU')
    reader = csv.reader(iFile, delimiter=',')

    for row in reader:
        if len(row[0]) > 3 and len(row[0]) <= 10:
            keyword = row[0].encode('utf-8')
            keywords.add(keyword)

    maxlen = max([len(k) for k in keywords])
    iFile.close()

    return None


# find longest matching suffix using a Bloom filter
def first_match(address):

    for i in range(maxlen, -1, -1):
        if address[-i:] in keywords:
            return address[-i:]

    # else return none
    return None


# simple exponential average
class exp_avg():
    r0 = None
    def smooth(self, r, prec = 1, alpha = 0.05):
        self.r0 = r if self.r0 == None else  alpha * r + (1 - alpha) * self.r0
        return round(self.r0, prec)


# main function
if __name__ == '__main__':

    # load interesting keywords
    load_keywords('names/NAMES.txt')
    print('Loaded {} keywords.'.format(len(keywords)))

    # connect to MongoDB (also check for failures)
    cred = os.getenv("MONGO_PASS")
    client = MongoClient('mongodb://' + ('' if cred is None else cred + '@') + 'localhost:27017/',
                         document_class=OrderedDict)
    try:
        client.admin.command('ismaster')
    except ConnectionFailure:
        print("MongoDB server not available!")
        sys.exit(-1)
    col = client.mydb.StellarVanityPool
    try:
        col.find_one({})
    except OperationFailure:
        print("Wrong MongoDB credentials provided!")
        sys.exit(-1)

    # generate!
    print('Existing number of vanity addresses: {}'.format(col.count('{}')))
    print('Generating new ones...')
    i = hits = oldhits = 0
    eavg = { "total": exp_avg(), "hit": exp_avg() }
    oldTime = time.perf_counter()

    while True:

        try:

            # generate and check for hit
            kp = Keypair.random()
            address = kp.address()
            hit = first_match(address)

            # insert hit into database
            if hit is not None:
                hits = hits + 1
                col.insert_one(OrderedDict([("suffix", hit.decode()),
                                            ("public", address.decode()),
                                            ("secret", kp.seed().decode())]))

            # display performance stats
            i = i + 1
            if i % NBENCH == 0:
                newTime = time.perf_counter()
                dt = newTime - oldTime
                print('Generation rate:', eavg["total"].smooth(1.0 * NBENCH / dt), 'addr/sec\t',
                      'Hit rate: ',  eavg["hit"].smooth((hits - oldhits) / dt), 'addr/sec')
                oldTime = newTime
                oldhits = hits

        # exit nicely if interrupted
        except KeyboardInterrupt:

            client.close()
            print('\nBye.')
            sys.exit(0)
